src/CoreBundle/Controller/ChameleonController.php

```diff
@@ -238,22 +240,27 @@ protected function sendDefaultHeaders()
      */
     protected function GeneratePage($pagedef)
     {
-        $pagedefData = $this->getPagedefData($pagedef);
-        if (false === $pagedefData) {
+        $pagedefData = $this->dataAccessCmsMasterPagedef->get($pagedef);
+        if (null === $pagedefData) {
             return new Response('<div style="background-color: #ffcccc; color: #900; border: 2px solid #c00; padding-left: 10px; padding-right: 10px; padding-top: 5px; padding-bottom: 5px; font-weight: bold; font-size: 11px; min-height: 40px; display: block;">Error: invalid page definition: '.TGlobal::OutHTML($pagedef).'</div>', Response::HTTP_NOT_FOUND);
         }
 
-        $this->moduleLoader->LoadModules($pagedefData['moduleList']);
+        $this->moduleLoader->LoadModules($pagedefData->getModuleList());
 
         foreach ($this->moduleLoader->modules as $sSpotName => $module) {
             $this->moduleLoader->modules[$sSpotName]->InjectVirtualModuleSpots($this->moduleLoader);
         }
         reset($this->moduleLoader->modules);
 
         $this->InitializeModules();
+
+        if (false === $this->checkAccess($pagedefData)) {
+            return new Response('You have no sufficient rights to view this page.', Response::HTTP_FORBIDDEN);
```

Comment:
This is backend logic only. So IMHO this also needs to be moved into the BackendController

Reply:
It depends however on the 10 lines before it - and ChameleonBackendController has nothing of this yet.